### DISCUSSION TITLE
[Feature] DAG Stratification

### DISCUSSION THREAD

**Discussion by ghost** [networkx/networkx#4755] [Algorithms]
Hi, a very common operation I find myself performing on a DAG is "stratifying" it, such that each level contains *all* the nodes whose dependencies have been satisfied by the nodes in a previous level. What this translates to is, take all the nodes in the DAG that don't have any dependencies and put them in list. Then "remove" those nodes from the DAG, and repeat the process, creating a new list at each step. This is a pretty standard BFS style topological sort, with the addition of keeping track of the "layer" each node is in.

Example:
```
# Imagine all arrows pointing downwards in this DAG

  F   G
  |\ /|
D | E |
 \|/  |
  B   C
   \ /
    A

Result:
[[D, F, G],
 [C, E],
 [B],
 [A]]
```

Real-world Application:
This stratified representation of a graph is useful in building quick/naive parallel execution schemes of DAG organized tasks. You can simply traverse the levels sequentially and execute the tasks (nodes) in each level in parallel. Waiting for the level to finish before moving on to the next. While this does not give you the "optimal" runtime for parallel execution of a DAG, it has the benefit of not requiring any distributed event-loop/scheduler, and can be executed in a simple for loop/map.

Current Solution:
As far as I'm aware, there's currently no built-in way to do this in networkx. One way it can be done are to implement your own topological sort that keeps track of the levels. Another way to do it is to get the max-path-distance from any node to the farthest leaf level. This second solution is not trivial to do with existing networkx utilities. It requires creating a dummy node as an ancestor for all the current leaf nodes, and then calling nx.all_simple_paths for every node to the dummy node, and taking the length of the longest path for each node. At that point it would likely be better to just do a recursive max-path-sum solution, outside of the networkx ecosystem.

Proposal:
There are a few ways I believe this could be implemented. I've outlined one that I believe balances minimizing changes and reusing existing algorithm implementations (namely topological_sort).

1) Update `networkx.algorithms.dag.topological_sort` to provide a BFS-style topological sort. This can simply be done by replacing `zero_indegree` with a deque (using popleft), instead of a list (using pop). Also, keep track of the current BF depth, this will not require any changes to the core algorithm (or space/time complexity), only a couple trivial modifications.
2) Add an option to topological_sort called something like `with_depth=False`. This option will enable the node yielded from the generator to be a tuple of `(node, depth as determined by a BFS)`
3) Add a function called `topological_stratify`, which will simply call topological_sort with the depths enabled and yield lists, with nodes placed in the list based on the returned depth.

Alternative Proposals:
- Rather than adding a new public facing option to topological_sort, instead make topological_sort a class implementing `__call__` so that the public interface is the same, but it can keep state of the nodes and depths, which can then be accessed by topological_stratify.
- Or, just provide an independent solution, most likely using a recursive max-path-distance to leaf level, so as not to duplicate work already being done in other existing algorithms.

Considering this depth/level information is (1) an intrinsic property of the DAG, (2) already implicitly determined through other algorithms and (3) has real-world utility, I think that it would make sense for the networkx ecosystem to provide a way to directly extract that information.

Implementation of the main proposal describe above, can be found here:
https://github.com/ArtinSarraf/networkx/pull/1

**Comment by dschult** [comment]
This seems like a reasonable proposal. You could make it a PR from your version of the code and we'll start the review process.

My main concern is that there are many slightly different ways of declaring the depth/level of a node in a topological sort. This choice: "each level is all nodes that have zero in_degree when all previous level nodes are removed" is a reasonable choice.

But I believe a similar notion of level is "those nodes that can be reached from nodes on the previous level". This is like a BFS level starting from the set of nodes with zero in_degree. It could be obtained from a BFS I guess.

Are there other notions of "depth" we should be paying attention to?

Other questions:
- Will you proposed solution change the topological sort order? (if I understand, it changes the code from a DFS to a BFS)
- Could you use `.pop(0)` instead of `deque` and what are the implications of each?
- We should choose a word to describe this notion of "depth" and stick to it. I started using "level" here for some reason (sorry). Is "stratify" the right word?  Is there a "depth" related version of "stratify" better than "stratification layer"?   Perhaps "depth" is the best word in which case `topological_stratify` could become `topological_depth`?

Thanks for this!

**Reply by ghost** [reply]
Agreed about the ambiguity of the term depth. Even with the BFS-depth qualifier. I think a more appropriate term would be "__generation__". As this goes nicely with the existing concept of "__ancestors__" and "__descendants__" in the DAG. I don't think generation would be ambiguous because it implies that all ancestors must come in a generation before it, and all descendants in a generation after.

With this "generation" term in mind, perhaps a better function name than "topological_stratify" would be something along the lines of "__genealogy__" , "__delineate__", or maybe the most straightforward "__topological_generations__".

To address your bullets:
- Will this change the sort order? Yes. However, I don't think this is a problem, as there is not one true, topological sort order, (in fact in cases where there is, e.g. a simple linked list, this will produce the same result).
- Could you use `.pop(0)`? You could, however this is not efficient. As it's an O(n) operation, as opposed to a `deque.popleft()` which is an O(1) operation. Here is an article that profiles the 2 and outlines the differences (https://medium.com/@shuangzizuobh2/how-well-do-you-code-python-9bec36bbc322)
- New term to describe the notion of depth: "generation"

One other idea, is that we could switch to using the deque/BFS style topological sort only on `with_generations=True`, and keep the existing list/DFS style otherwise. Although I think this would be needlessly complex.

**Reply by ghost** [reply]
PR added here (with proposed generation changes made):
https://github.com/networkx/networkx/pull/4757
